sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,137 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp   session  state    td_name  elfpath
+ * 0       1       0       1       1       R       init    /bin/ksh
+ *
+ * XXX: Each instance of /dev/procstat has 4K bytes for process entries.
+ * It will be enough for about 40 processes (if we assume average size of entry
+ * to be 100 characters).
+ *
+ * Until we run more processes than ksh and some tests, this solution don't need
+ * to be more complex.
+ */
+
+#define PSTAT_SIZE 4096
+#define BUF_SIZE (PSTAT_SIZE - sizeof(pstat_t))
+
+static char state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct pstat_data {
+  size_t len;
+  char buf[];
+} pstat_t;
+
+static int dev_procstat_read(file_t *f, uio_t *uio) {
+  pstat_t *pst = f->f_data;
+
+  if ((size_t)f->f_offset >= pst->len)
+    return 0;
+
+  uio->uio_offset = f->f_offset;
+  size_t len = MIN(uio->uio_resid, pst->len - uio->uio_offset);
+  int error = uiomove(pst->buf, len, uio);
+
+  f->f_offset = uio->uio_offset;
+  return error;
+}
+
+static fileops_t dev_procstat_fileops = {
+  .fo_read = dev_procstat_read,
+  .fo_close = default_vnclose,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = default_vnioctl,
+};
+
+/* Write data to buffer and return number of written characters.
+ *
+ * We also have to handle 'proc0' here. It is on proc_list but we
+ * don't want to show it to user.
+ */
+static int write_to_buf(char *buf, size_t max_len, proc_t *p) {
+  if (p->p_pid == 0)
+    return 0;
+
+  SCOPED_MTX_LOCK(&p->p_lock);
+  return snprintf(buf, max_len, "%d\t%d\t%d\t%d\t%d\t%c\t%s\t%s\n",
```

Comment:
I wonder... why don't you:
 * devise a table of structures (let's say `procstat_t`),
 * fill it with data from `proc_t` (that's going to be really fast),
 * convert `procstat_t` to text when user reads the file.

It'll be an extra step, but you can easily estimate how much memory you need to store `procstat_t` array.

Reply:
It sounds good. There is one problem. When user calls read it can end inside one table entry. It will bring some weird logic, but I think it could be done.

Reply:
done